MoreLinq/Interleave.cs

```diff
@@ -45,125 +43,63 @@ public static partial class MoreEnumerable
-        /// <param name="sequence">The first sequence in the interleave group</param>
-        /// <param name="imbalanceStrategy">Defines the behavior of the operator when sequences are of unequal length</param>
-        /// <param name="otherSequences">The other sequences in the interleave group</param>
-        /// <returns>A sequence of interleaved elements from all of the source sequences</returns>
-
-        static IEnumerable<T> Interleave<T>(this IEnumerable<T> sequence, ImbalancedInterleaveStrategy imbalanceStrategy, params IEnumerable<T>[] otherSequences)
         {
             if (sequence == null) throw new ArgumentNullException(nameof(sequence));
             if (otherSequences == null) throw new ArgumentNullException(nameof(otherSequences));
-            if (otherSequences.Any(s => s == null))
-                throw new ArgumentNullException(nameof(otherSequences), "One or more sequences passed to Interleave was null.");
 
-            return _(); IEnumerable<T> _()
-            {
-                var sequences = new[] { sequence }.Concat(otherSequences);
+            return InterleaveSkip(otherSequences.Prepend(sequence));
+        }
 
-                // produce an iterator collection for all IEnumerable<T> instancess passed to us
-                var iterators = sequences.Select(e => e.GetEnumerator()).Acquire();
-                List<IEnumerator<T>> iteratorList = null;
+        private static IEnumerable<T> InterleaveSkip<T>(this IEnumerable<IEnumerable<T>> sequences)
+        {
+            var enumerators = new LinkedList<IEnumerator<T>>();
 
-                try
+            try
+            {
+                // First pass. create enumerators.
+                foreach (var sequence in sequences)
                 {
-                    iteratorList = new List<IEnumerator<T>>(iterators);
-                    iterators = null;
-                    var shouldContinue = true;
-                    var consumedIterators = 0;
-                    var iterCount = iteratorList.Count;
-
-                    while (shouldContinue)
-                    {
-                        // advance every iterator and verify a value exists to be yielded
-                        for (var index = 0; index < iterCount; index++)
-                        {
-                            if (!iteratorList[index].MoveNext())
-                            {
-                                // check if all iterators have been consumed and we can terminate
-                                // or if the imbalance strategy informs us that we MUST terminate
-                                if (++consumedIterators == iterCount || imbalanceStrategy == ImbalancedInterleaveStrategy.Stop)
-                                {
-                                    shouldContinue = false;
-                                    break;
-                                }
+                    if (sequence == null)
```

Comment:
This argument check will take place when the iteration starts and not when `Interleave` is called. It would need to be moved out.

Reply:
Indeed. #697 drives me to this inadequate implementation.

Reply:
By the way a test for that is missing here.